src/execution/compiler/operator/index_create_translator.cpp

```diff
@@ -36,71 +35,109 @@ IndexCreateTranslator::IndexCreateTranslator(const planner::CreateIndexPlanNode
     compilation_context->Prepare(*index_col.StoredExpression());
   }
   pipeline->RegisterSource(this, Pipeline::Parallelism::Serial);
+  // col_oids is a global array
+  ast::Expr *arr_type = codegen_->ArrayType(all_oids_.size(), ast::BuiltinType::Kind::Uint32);
+  global_col_oids_ = compilation_context->GetQueryState()->DeclareStateEntry(codegen_, "global_col_oids", arr_type);
+  // storage interface is local to pipeline
+  ast::Expr *storage_interface_type = codegen_->BuiltinType(ast::BuiltinType::StorageInterface);
+  local_storage_interface_ = pipeline->DeclarePipelineStateEntry("local_storage_interface", storage_interface_type);
+  // index pr is local to pipeline
+  ast::Expr *index_pr_type = codegen_->BuiltinType(ast::BuiltinType::ProjectedRow);
+  local_index_pr_ = pipeline->DeclarePipelineStateEntry("local_index_pr", codegen_->PointerType(index_pr_type));
+  // tuple slot is local to pipeline
+  // TODO(wuwenw): do we really need a local copy of tuple slot?
```

Comment:
You seem to be using it in `if (!@indexInsertWithSlot(&pipelineState.local_storage_interface, &pipelineState.local_tuple_slot, false))`

Reply:
Yeah I mean if there is a way to directly insert from vpi, instead of making local copy.